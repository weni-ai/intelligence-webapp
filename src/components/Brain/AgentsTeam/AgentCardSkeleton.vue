<template>
  <section class="agent-card-skeleton">
    <UnnnicSkeletonLoading
      tag="div"
      width="30%"
      height="22px"
    />

    <UnnnicSkeletonLoading
      tag="div"
      width="100%"
      height="44px"
    />

    <section class="agent-card-skeleton__skills">
      <UnnnicSkeletonLoading
        tag="div"
        width="100px"
        height="30px"
      />
      <UnnnicSkeletonLoading
        tag="div"
        width="150px"
        height="30px"
      />
    </section>

    <UnnnicSkeletonLoading
      tag="div"
      width="100%"
      height="38px"
    />
  </section>
</template>

<style lang="scss" scoped>
.agent-card-skeleton {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-xs;

  &__skills {
    width: 100%;
    display: flex;
    gap: $unnnic-spacing-nano;
    flex-wrap: wrap;
  }
}
</style>
