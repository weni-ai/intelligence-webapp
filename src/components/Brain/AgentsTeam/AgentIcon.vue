<template>
  <img
    v-if="icon && agentIcon"
    :src="agentIcon"
    class="content__icon"
    data-testid="agent-icon"
  />
</template>

<script setup>
import { computed } from 'vue';
import { agentIcons } from '@/utils/agentIconService';

const props = defineProps({
  icon: {
    type: String,
    default: '',
  },
});

const agentIcon = computed(() => {
  if (!props.icon) return;

  return agentIcons[props.icon] || null;
});
</script>
