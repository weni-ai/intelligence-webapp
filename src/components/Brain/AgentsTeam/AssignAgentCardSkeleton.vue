<template>
  <section
    class="assign-agent-card-skeleton"
    data-testid="assign-agent-card-skeleton"
  >
    <UnnnicSkeletonLoading
      tag="div"
      width="30%"
      height="22px"
      data-testid="skeleton-loading"
    />

    <UnnnicSkeletonLoading
      tag="div"
      width="100%"
      height="44px"
      data-testid="skeleton-loading"
    />

    <section
      class="assign-agent-card-skeleton__skills"
      data-testid="skills-section"
    >
      <UnnnicSkeletonLoading
        tag="div"
        width="100px"
        height="30px"
        data-testid="skills-skeleton-loading"
      />
      <UnnnicSkeletonLoading
        tag="div"
        width="150px"
        height="30px"
        data-testid="skills-skeleton-loading"
      />
    </section>

    <UnnnicSkeletonLoading
      tag="div"
      width="100%"
      height="38px"
      data-testid="skeleton-loading"
    />
  </section>
</template>

<style lang="scss" scoped>
.assign-agent-card-skeleton {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-xs;

  &__skills {
    width: 100%;
    display: flex;
    gap: $unnnic-spacing-nano;
    flex-wrap: wrap;
  }
}
</style>
