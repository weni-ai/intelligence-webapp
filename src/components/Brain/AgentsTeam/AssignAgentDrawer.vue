<template>
  <UnnnicDrawer
    v-if="agent.credentials?.length"
    :modelValue="modelValue"
    class="assign-agent-drawer"
    :title="$t('router.agents_team.drawer.title', { name: agent.name })"
    :primaryButtonText="$t('router.agents_team.drawer.assign')"
    :disabledPrimaryButton="!hasAllCredentialsWithValue"
    :loadingPrimaryButton="
      tuningsStore.credentials.status === 'loading' || isAssigning
    "
    data-testid="assign-agent-drawer"
    @primary-button-click="toggleAgentAssignment"
    @close="close"
  >
    <template #content>
      <section class="assign-agent-drawer__content">
        <section
          v-if="hasCredentialsWithValue"
          class="assign-agent-drawer__credentials"
          data-testid="assign-agent-drawer-credentials"
        >
          <UnnnicIntelligenceText
            tag="p"
            family="secondary"
            size="body-gt"
            color="neutral-dark"
          >
            {{ $t('router.agents_team.drawer.credentials_used_by_this_agent') }}
          </UnnnicIntelligenceText>

          <section class="credentials__list">
            <UnnnicIntelligenceText
              v-for="(credential, key) in credentialsWithValue"
              :key="key"
              tag="p"
              family="secondary"
              size="body-md"
              color="neutral-clean"
            >
              {{ credential.label }}
            </UnnnicIntelligenceText>
          </section>
        </section>

        <template v-if="credentialsWithoutValue.length">
          <UnnnicIntelligenceText
            tag="p"
            family="secondary"
            size="body-gt"
            color="neutral-cloudy"
          >
            {{ $t('router.agents_team.drawer.description') }}
          </UnnnicIntelligenceText>

          <UnnnicFormElement
            v-for="(credential, key) in credentialsWithoutValue"
            :key="key"
            :label="credential.label"
            data-testid="assign-agent-drawer-form-element"
          >
            <UnnnicInput
              :modelValue="getCredentialValue(credential.name)"
              :placeholder="credential.placeholder || credential.label"
              :nativeType="credential.is_confidential ? 'password' : 'text'"
              data-testid="assign-agent-drawer-input"
              @update:model-value="
                tuningsStore.updateCredential({
                  ...credential,
                  value: $event,
                })
              "
            />
          </UnnnicFormElement>

          <section
            class="assign-agent-drawer__shared-info"
            data-testid="assign-agent-drawer-shared-info"
          >
            <UnnnicIcon
              scheme="neutral-cleanest"
              icon="info"
              size="sm"
              filled
              data-testid="assign-agent-drawer-shared-info-icon"
            />
            <UnnnicIntelligenceText
              tag="p"
              family="secondary"
              size="body-md"
              color="neutral-clean"
              data-testid="assign-agent-drawer-shared-info-text"
            >
              {{
                $t(
                  'router.agents_team.drawer.credentials_shared_with_all_agents',
                )
              }}
            </UnnnicIntelligenceText>
          </section>
        </template>
      </section>
    </template>
  </UnnnicDrawer>
</template>

<script setup>
import { useTuningsStore } from '@/store/Tunings';
import { computed } from 'vue';

const props = defineProps({
  modelValue: {
    type: Boolean,
    required: true,
  },
  agent: {
    type: Object,
    required: true,
  },
  isAssigning: {
    type: Boolean,
    required: false,
  },
});

const emit = defineEmits(['update:modelValue', 'assign']);

const tuningsStore = useTuningsStore();

const hasCredentialsWithValue = computed(() => {
  return props.agent.credentials?.some((credential) => {
    const [index, type] = tuningsStore.getCredentialIndex(credential.name);

    return tuningsStore.initialCredentials?.[type]?.[index]?.value;
  });
});

const credentialsWithoutValue = computed(() => {
  return (
    props.agent.credentials?.filter((credential) => {
      const [index, type] = tuningsStore.getCredentialIndex(credential.name);
      const value = tuningsStore.initialCredentials?.[type]?.[index]?.value;
      return !value?.trim();
    }) || []
  );
});

const credentialsWithValue = computed(() => {
  const withoutValueNames = credentialsWithoutValue?.value.map(
    (cred) => cred.name,
  );
  return (
    props.agent.credentials?.filter(
      (credential) =>
        getCredentialValue(credential.name) &&
        !withoutValueNames.includes(credential.name),
    ) || []
  );
});

const hasAllCredentialsWithValue = computed(() => {
  return props.agent.credentials?.every((credential) => {
    const [index, type] = tuningsStore.getCredentialIndex(credential.name);
    const value = tuningsStore.credentials.data?.[type]?.[index]?.value;
    return !!value?.trim();
  });
});

function close() {
  emit('update:modelValue', false);
}

async function toggleAgentAssignment() {
  try {
    if (credentialsWithoutValue.value.length) {
      const credentialsToCreate = credentialsWithoutValue.value.map(
        (credential) => {
          const [index, type] = tuningsStore.getCredentialIndex(
            credential.name,
          );
          return {
            ...credential,
            value: tuningsStore.credentials.data?.[type]?.[index]?.value,
          };
        },
      );
      await tuningsStore.createCredentials(
        props.agent.uuid,
        credentialsToCreate,
      );
    }

    emit('assign');
  } catch (error) {
    console.error(error);
  }
}

const getCredentialValue = (credentialName) => {
  const [index, type] = tuningsStore.getCredentialIndex(credentialName);

  return tuningsStore.credentials.data?.[type]?.[index]?.value;
};
</script>

<style lang="scss" scoped>
.assign-agent-drawer {
  &__content {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-sm;
  }

  &__credentials {
    border-radius: $unnnic-border-radius-md;
    border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

    padding: $unnnic-spacing-sm;

    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-nano;

    .credentials__list {
      display: flex;
      flex-direction: column;
    }
  }

  &__shared-info {
    border-radius: $unnnic-border-radius-sm;
    border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

    padding: $unnnic-spacing-ant;

    display: flex;
    align-items: center;
    gap: $unnnic-spacing-xs;
  }
}
</style>
