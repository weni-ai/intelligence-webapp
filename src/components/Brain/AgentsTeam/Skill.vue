<template>
  <section class="skill">
    <p
      class="skill__icon"
      data-testid="skill-icon"
    >
      {{ icon }}
    </p>

    <UnnnicIntelligenceText
      class="skill__title"
      tag="p"
      family="secondary"
      size="body-md"
      color="neutral-cloudy"
      data-testid="skill-name"
    >
      {{ title }}
    </UnnnicIntelligenceText>
  </section>
</template>

<script setup>
const props = defineProps({
  title: {
    type: String,
    required: true,
  },
  icon: {
    type: String,
    required: true,
  },
});
</script>

<style lang="scss" scoped>
.skill {
  border-radius: $unnnic-border-radius-sm;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

  padding: $unnnic-spacing-nano $unnnic-spacing-xs;

  display: flex;
  align-items: center;
  gap: $unnnic-spacing-nano;

  &__icon {
    font-size: $unnnic-icon-size-xs;
    line-height: 0;
  }

  &__title {
    white-space: nowrap;
  }
}
</style>
