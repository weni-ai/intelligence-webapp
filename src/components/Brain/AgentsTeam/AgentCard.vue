<template>
  <section
    data-testid="agent-card"
    :class="['agent-card', { 'agent-card--empty': empty }]"
  >
    <AgentCardSkeleton
      v-if="loading"
      data-testid="agent-card-skeleton"
    />

    <AgentCardEmpty
      v-else-if="empty"
      data-testid="agent-card-empty"
    />

    <section
      v-else
      class="agent-card__content"
    >
      <UnnnicIntelligenceText
        tag="p"
        family="secondary"
        size="body-gt"
        color="neutral-darkest"
        weight="bold"
        data-testid="title"
      >
        {{ title }}
      </UnnnicIntelligenceText>

      <UnnnicIntelligenceText
        v-if="description"
        tag="p"
        family="secondary"
        size="body-gt"
        color="neutral-cloudy"
        data-testid="description"
      >
        {{ description }}
      </UnnnicIntelligenceText>

      <section class="content__skills">
        <Skill
          v-for="skill in skills"
          :key="skill.name"
          :title="skill.name"
          :icon="skill.icon"
          data-testid="skill"
        />
      </section>
    </section>

    <UnnnicButton
      v-if="!loading && !empty && assignment"
      :class="[
        'agent-card__button',
        { 'agent-card__button--assigned': assigned },
      ]"
      :text="
        assigned
          ? $t('router.agents_team.card.assigned')
          : $t('router.agents_team.card.assign')
      "
      :type="assigned ? 'tertiary' : 'primary'"
      :iconLeft="assigned ? 'check' : ''"
      size="small"
      :loading="isAssigning"
      data-testid="assign-button"
      @click="toggleAgentAssignment"
    />
  </section>
</template>

<script setup>
import { ref } from 'vue';

import { useAgentsTeamStore } from '@/store/AgentsTeam';

import AgentCardSkeleton from './AgentCardSkeleton.vue';
import AgentCardEmpty from './AgentCardEmpty.vue';
import Skill from './Skill.vue';

defineEmits(['assign']);

const props = defineProps({
  loading: {
    type: Boolean,
    default: false,
  },
  title: {
    type: String,
    default: '',
  },
  description: {
    type: String,
    default: '',
  },
  /**
   * An array of skills where each skill is an object with the following properties:
   * @property {string} name - The name of the skill
   * @property {string} icon - The icon of the skill
   */
  skills: {
    type: Array,
    default: () => [],
    validator(skills) {
      return skills.every((skill) => 'name' in skill && 'icon' in skill);
    },
  },
  assignment: {
    type: Boolean,
    default: true,
  },
  assigned: {
    type: Boolean,
    default: false,
  },
  uuid: {
    type: String,
    default: '',
  },
  empty: {
    type: Boolean,
    default: false,
  },
});

const isAssigning = ref(false);
const agentsTeamStore = useAgentsTeamStore();

async function toggleAgentAssignment() {
  isAssigning.value = true;

  try {
    await agentsTeamStore.toggleAgentAssignment({
      uuid: props.uuid,
      is_assigned: !props.assigned,
    });
  } catch (error) {
    console.error(error);
  } finally {
    isAssigning.value = false;
  }
}
</script>

<style lang="scss" scoped>
.agent-card {
  border-radius: $unnnic-border-radius-md;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-cleanest;

  padding: $unnnic-spacing-sm;

  display: grid;
  gap: $unnnic-spacing-sm;
  align-content: space-between;

  &--empty {
    display: flex;

    cursor: pointer;
  }

  &__content {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-xs;

    .content__skills {
      display: flex;
      gap: $unnnic-spacing-nano;
      flex-wrap: wrap;
    }
  }

  :deep(.unnnic-button).agent-card__button {
    &--assigned {
      color: $unnnic-color-weni-600;
    }

    [class*='icon'] {
      color: $unnnic-color-weni-600;
    }
  }
}
</style>
