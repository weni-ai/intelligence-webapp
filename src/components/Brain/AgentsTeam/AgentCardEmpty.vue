<template>
  <button
    class="agent-card-empty"
    data-testid="agent-card-empty"
    @click="createNewAgent"
  >
    <UnnnicIcon
      icon="add"
      size="md"
      scheme="neutral-clean"
      data-testid="agent-card-empty-icon"
    />

    <UnnnicIntelligenceText
      tag="p"
      family="secondary"
      size="body-gt"
      color="neutral-dark"
      data-testid="agent-card-empty-title"
    >
      {{ $t('router.agents_team.create_agent') }}
    </UnnnicIntelligenceText>
  </button>
</template>

<script setup>
import { useAgentsTeamStore } from '@/store/AgentsTeam';

const agentsTeamStore = useAgentsTeamStore();
function createNewAgent() {
  window.open(agentsTeamStore.linkToCreateAgent, '_blank');
}
</script>

<style lang="scss" scoped>
.agent-card-empty {
  border: none;
  background: none;

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  height: 100%;
  width: 100%;

  cursor: pointer;
}
</style>
