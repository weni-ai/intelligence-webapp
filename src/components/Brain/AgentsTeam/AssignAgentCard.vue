<template>
  <section
    data-testid="assign-agent-card"
    :class="['assign-agent-card', { 'assign-agent-card--empty': empty }]"
  >
    <AssignAgentCardSkeleton
      v-if="loading"
      data-testid="assign-agent-card-skeleton"
    />

    <AssignAgentCardEmpty
      v-else-if="empty"
      data-testid="assign-agent-card-empty"
    />

    <section
      v-else
      class="assign-agent-card__content"
    >
      <UnnnicIntelligenceText
        tag="p"
        family="secondary"
        size="body-gt"
        color="neutral-darkest"
        weight="bold"
        data-testid="title"
      >
        {{ agent.name }}
      </UnnnicIntelligenceText>

      <UnnnicIntelligenceText
        v-if="agent.description && assignment"
        class="content__description"
        tag="p"
        family="secondary"
        size="body-gt"
        color="neutral-cloudy"
        data-testid="description"
      >
        {{ agent.description }}
      </UnnnicIntelligenceText>

      <section class="content__skills">
        <Skill
          v-for="skill in agent.skills"
          :key="skill.name"
          :title="skill.name"
          :icon="skill.icon"
          data-testid="skill"
        />
      </section>
    </section>

    <UnnnicButton
      v-if="!loading && !empty && assignment"
      :class="[
        'assign-agent-card__button',
        { 'assign-agent-card__button--assigned': agent.assigned },
      ]"
      :text="
        agent.assigned
          ? $t('router.agents_team.card.assigned')
          : $t('router.agents_team.card.assign')
      "
      :type="agent.assigned ? 'secondary' : 'primary'"
      :iconLeft="agent.assigned ? 'check' : ''"
      size="small"
      :loading="isAssigning"
      data-testid="assign-button"
      @click="
        !agent.assigned && agent.credentials?.length > 0
          ? toggleDrawer()
          : toggleAgentAssignment()
      "
    />
  </section>

  <AssignAgentDrawer
    v-model="isAssignDrawerOpen"
    :agent="agent"
    :isAssigning="isDrawerAssigning"
    @assign="toggleDrawerAssigning"
  />
</template>

<script setup>
import { ref } from 'vue';

import { useAgentsTeamStore } from '@/store/AgentsTeam';

import AssignAgentCardSkeleton from './AssignAgentCardSkeleton.vue';
import AssignAgentCardEmpty from './AssignAgentCardEmpty.vue';
import Skill from './Skill.vue';
import AssignAgentDrawer from './AssignAgentDrawer.vue';

const props = defineProps({
  loading: {
    type: Boolean,
    default: false,
  },
  agent: {
    type: Object,
    default: () => ({}),
  },
  assignment: {
    type: Boolean,
    default: true,
  },
  empty: {
    type: Boolean,
    default: false,
  },
});

const agentsTeamStore = useAgentsTeamStore();

const isAssignDrawerOpen = ref(false);
const isAssigning = ref(false);
const isDrawerAssigning = ref(false);

async function toggleDrawer() {
  isAssignDrawerOpen.value = !isAssignDrawerOpen.value;
}

async function assignAgent() {
  try {
    await agentsTeamStore.toggleAgentAssignment({
      uuid: props.agent.uuid,
      is_assigned: !props.agent.assigned,
    });
  } catch (error) {
    console.error(error);
  }
}

async function toggleAgentAssignment() {
  isAssigning.value = true;
  await assignAgent();
  isAssigning.value = false;
}

async function toggleDrawerAssigning() {
  isDrawerAssigning.value = true;
  await assignAgent();
  isDrawerAssigning.value = false;
  toggleDrawer();
}
</script>

<style lang="scss" scoped>
.assign-agent-card {
  border-radius: $unnnic-border-radius-md;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-cleanest;

  padding: $unnnic-spacing-sm;

  display: grid;
  gap: $unnnic-spacing-sm;
  align-content: space-between;

  &--empty {
    display: flex;

    cursor: pointer;
  }

  &__content {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-xs;

    .content__description {
      display: -webkit-box;
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;
      overflow: hidden;
    }

    .content__skills {
      display: flex;
      gap: $unnnic-spacing-nano;
      flex-wrap: wrap;
    }
  }

  :deep(.unnnic-button).assign-agent-card__button {
    &--assigned {
      color: $unnnic-color-weni-600;
    }

    [class*='icon'] {
      color: $unnnic-color-weni-600;
    }
  }
}
</style>
