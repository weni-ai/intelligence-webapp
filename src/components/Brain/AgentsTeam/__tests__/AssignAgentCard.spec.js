import { afterEach, beforeEach, it, vi } from 'vitest';
import { mount } from '@vue/test-utils';

import { createTestingPinia } from '@pinia/testing';
import { useAgentsTeamStore } from '@/store/AgentsTeam';
import { useTuningsStore } from '@/store/Tunings';
import i18n from '@/utils/plugins/i18n';

import AssignAgentCard from '../AssignAgentCard.vue';

const pinia = createTestingPinia({
  initialState: {
    officialAgents: {
      status: null,
      data: [],
    },
    myAgents: {
      status: null,
      data: [],
    },
  },
});

describe('AssignAgentCard.vue', () => {
  let wrapper;

  const agentsTeamStore = useAgentsTeamStore();
  const tuningsStore = useTuningsStore();

  beforeEach(() => {
    wrapper = mount(AssignAgentCard, {
      global: {
        plugins: [pinia],
        stubs: {
          AssignAgentDrawer: true,
        },
      },
      props: {
        loading: false,
        agent: {
          name: 'Test Title',
          description: 'Test Description',
          skills: [
            { name: 'Skill 1', icon: 'icon-1' },
            { name: 'Skill 2', icon: 'icon-2' },
          ],
        },
      },
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should renders correctly when loading', async () => {
    await wrapper.setProps({ loading: true });

    expect(
      wrapper
        .findComponent('[data-testid="assign-agent-card-skeleton"]')
        .exists(),
    ).toBe(true);
  });

  it('should renders correctly when not loading', () => {
    expect(wrapper.find('[data-testid="title"]').text()).toBe('Test Title');
    expect(wrapper.find('[data-testid="description"]').text()).toBe(
      'Test Description',
    );
    expect(wrapper.findAll('[data-testid="skill"]').length).toBe(2);
  });

  it('should renders skills correctly', () => {
    const skills = wrapper.findAllComponents('[data-testid="skill"]');
    expect(skills[0].props('title')).toBe('Skill 1');
    expect(skills[0].props('icon')).toBe('icon-1');
    expect(skills[1].props('title')).toBe('Skill 2');
    expect(skills[1].props('icon')).toBe('icon-2');
  });

  describe('ContentHeader', () => {
    it('should render header actions when assignment is false', async () => {
      await wrapper.setProps({ assignment: false });
      expect(
        wrapper.find('[data-testid="assign-agent-card-actions"]').exists(),
      ).toBe(true);
    });

    it('should not render header actions when assignment is true', () => {
      expect(
        wrapper.find('[data-testid="assign-agent-card-actions"]').exists(),
      ).toBe(false);
    });

    it('should render official tag when agent is official', async () => {
      await wrapper.setProps({
        assignment: false,
        agent: {
          ...wrapper.props('agent'),
          is_official: true,
        },
      });

      const tag = wrapper.findComponent('[data-testid="agent-tag"]');
      expect(tag.props('text')).toBe(
        i18n.global.t('router.agents_team.card.official'),
      );

      expect(tag.props('scheme')).toBe('weni');
    });

    it('should render custom tag when agent is not official', async () => {
      await wrapper.setProps({
        assignment: false,
        agent: {
          ...wrapper.props('agent'),
          is_official: false,
        },
      });

      const tag = wrapper.findComponent('[data-testid="agent-tag"]');
      expect(tag.props('text')).toBe(
        i18n.global.t('router.agents_team.card.custom'),
      );
      expect(tag.props('scheme')).toBe('aux-purple');
    });
  });

  describe('Assign button', () => {
    beforeEach(async () => {
      await wrapper.setProps({ assignment: true });
    });

    const assignButton = () =>
      wrapper.findComponent('[data-testid="assign-button"]');

    test('renders assign button with correct type and iconLeft', async () => {
      expect(assignButton().props('type')).toBe('primary');
      expect(assignButton().props('iconLeft')).toBe('');

      await wrapper.setProps({
        agent: { ...wrapper.props('agent'), assigned: true },
      });

      expect(assignButton().props('type')).toBe('secondary');
      expect(assignButton().props('iconLeft')).toBe('check');
    });

    it('calls toggleAgentAssignment when button is clicked for an agent without credentials', async () => {
      await wrapper.setProps({
        agent: { ...wrapper.props('agent'), assigned: false },
      });

      const toggleAgentAssignmentSpy = vi.spyOn(
        wrapper.vm,
        'toggleAgentAssignment',
      );

      await assignButton().trigger('click');

      expect(toggleAgentAssignmentSpy).toHaveBeenCalled();
    });

    it('should log error when toggleAgentAssignment throws an error', async () => {
      const consoleErrorSpy = vi
        .spyOn(console, 'error')
        .mockImplementation(() => {});
      agentsTeamStore.toggleAgentAssignment = vi
        .fn()
        .mockRejectedValue(new Error(''));

      await wrapper.vm.toggleAgentAssignment();

      expect(consoleErrorSpy).toHaveBeenCalled();
    });

    describe('toggleDrawer', () => {
      it('should toggle the isAssignDrawerOpen when toggleDrawer called', async () => {
        expect(wrapper.vm.isAssignDrawerOpen).toBe(false);

        await wrapper.vm.toggleDrawer();

        expect(wrapper.vm.isAssignDrawerOpen).toBe(true);

        await wrapper.vm.toggleDrawer();

        expect(wrapper.vm.isAssignDrawerOpen).toBe(false);
      });

      it('should call toggleDrawer when clicking assign button with agent having credentials', async () => {
        await wrapper.setProps({
          agent: {
            ...wrapper.props('agent'),
            assigned: false,
            credentials: [{ id: 1 }],
          },
        });

        const toggleDrawerSpy = vi.spyOn(wrapper.vm, 'toggleDrawer');

        await assignButton().trigger('click');

        expect(toggleDrawerSpy).toHaveBeenCalled();
      });

      it('should not call toggleDrawer for agent without credentials', async () => {
        await wrapper.setProps({
          agent: {
            ...wrapper.props('agent'),
            assigned: false,
            credentials: [],
          },
        });

        const toggleDrawerSpy = vi.spyOn(wrapper.vm, 'toggleDrawer');
        const toggleAgentAssignmentSpy = vi.spyOn(
          wrapper.vm,
          'toggleAgentAssignment',
        );

        await assignButton().trigger('click');

        expect(toggleDrawerSpy).not.toHaveBeenCalled();
        expect(toggleAgentAssignmentSpy).toHaveBeenCalled();
      });

      it('should not call toggleDrawer for already assigned agent', async () => {
        await wrapper.setProps({
          agent: {
            ...wrapper.props('agent'),
            assigned: true,
            credentials: [{ id: 1 }],
          },
        });

        const toggleDrawerSpy = vi.spyOn(wrapper.vm, 'toggleDrawer');
        const toggleAgentAssignmentSpy = vi.spyOn(
          wrapper.vm,
          'toggleAgentAssignment',
        );

        await assignButton().trigger('click');

        expect(toggleDrawerSpy).not.toHaveBeenCalled();
        expect(toggleAgentAssignmentSpy).toHaveBeenCalled();
      });
    });
  });

  describe('assignAgent', () => {
    beforeEach(() => {
      agentsTeamStore.toggleAgentAssignment = vi.fn().mockResolvedValue({
        status: 'success',
      });

      tuningsStore.fetchCredentials = vi.fn().mockResolvedValue();
    });

    it('should emit agent-assigned event when agent is successfully assigned', async () => {
      await wrapper.setProps({
        assignment: true,
        agent: {
          ...wrapper.props('agent'),
          assigned: false,
          external_id: 'test-id',
        },
      });

      await wrapper.vm.assignAgent();

      expect(wrapper.emitted('agent-assigned')).toBeTruthy();
    });

    it('should not emit agent-assigned event when agent is successfully unassigned', async () => {
      await wrapper.setProps({
        assignment: true,
        agent: {
          ...wrapper.props('agent'),
          assigned: true,
          external_id: 'test-id',
        },
      });

      await wrapper.vm.assignAgent();

      expect(wrapper.emitted('agent-assigned')).toBeFalsy();
    });

    it('should call fetchCredentials when agent has credentials', async () => {
      await wrapper.setProps({
        assignment: true,
        agent: {
          ...wrapper.props('agent'),
          credentials: [{ id: 1 }],
          external_id: 'test-id',
        },
      });

      await wrapper.vm.assignAgent();

      expect(tuningsStore.fetchCredentials).toHaveBeenCalled();
    });

    it('should call fetchCredentials when assignment is false', async () => {
      await wrapper.setProps({
        assignment: false,
        agent: {
          ...wrapper.props('agent'),
          credentials: [],
          external_id: 'test-id',
        },
      });

      await wrapper.vm.assignAgent();

      expect(tuningsStore.fetchCredentials).toHaveBeenCalled();
    });

    it('should not call fetchCredentials when agent has no credentials and assignment is true', async () => {
      await wrapper.setProps({
        assignment: true,
        agent: {
          ...wrapper.props('agent'),
          credentials: [],
          external_id: 'test-id',
        },
      });

      await wrapper.vm.assignAgent();

      expect(tuningsStore.fetchCredentials).not.toHaveBeenCalled();
    });

    it('should not process success actions when status is not success', async () => {
      agentsTeamStore.toggleAgentAssignment = vi.fn().mockResolvedValue({
        status: 'error',
      });

      await wrapper.setProps({
        assignment: true,
        agent: {
          ...wrapper.props('agent'),
          assigned: false,
          credentials: [{ id: 1 }],
          external_id: 'test-id',
        },
      });

      await wrapper.vm.assignAgent();

      expect(wrapper.emitted('agent-assigned')).toBeFalsy();
      expect(tuningsStore.fetchCredentials).not.toHaveBeenCalled();
    });
  });

  describe('toggleDrawerAssigning function', () => {
    beforeEach(() => {
      vi.spyOn(wrapper.vm, 'assignAgent').mockResolvedValue();
      vi.spyOn(wrapper.vm, 'toggleDrawer').mockResolvedValue();
    });

    it('should set isDrawerAssigning to true during assignment process', async () => {
      expect(wrapper.vm.isDrawerAssigning).toBe(false);

      const process = wrapper.vm.toggleDrawerAssigning();

      expect(wrapper.vm.isDrawerAssigning).toBe(true);

      await process;
    });

    it('should set isDrawerAssigning to false after assignment process', async () => {
      await wrapper.vm.toggleDrawerAssigning();

      expect(wrapper.vm.isDrawerAssigning).toBe(false);
    });

    it('should assign agent during the process', async () => {
      await wrapper.vm.toggleDrawerAssigning();

      expect(agentsTeamStore.toggleAgentAssignment).toHaveBeenCalled();
    });

    it('should toggle drawer after the assignment process', async () => {
      wrapper.vm.isAssignDrawerOpen = true;

      await wrapper.vm.toggleDrawerAssigning();

      expect(wrapper.vm.isAssignDrawerOpen).toBe(false);
    });

    it('should still call toggleDrawer even if assignAgent throws an error', async () => {
      wrapper.vm.isAssignDrawerOpen = true;
      vi.spyOn(wrapper.vm, 'assignAgent').mockRejectedValue(new Error(''));

      await wrapper.vm.toggleDrawerAssigning();

      expect(wrapper.vm.isAssignDrawerOpen).toBe(false);
    });
  });
});
