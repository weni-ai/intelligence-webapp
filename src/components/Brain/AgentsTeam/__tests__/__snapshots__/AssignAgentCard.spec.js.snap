// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AssignAgentCard.vue > renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <section
    class="assign-agent-card"
    data-testid="assign-agent-card"
    data-v-73104bd4=""
  >
    <section
      class="assign-agent-card__content"
      data-v-73104bd4=""
    >
      <!--v-if-->
      <header
        class="content__header"
        data-v-73104bd4=""
      >
        <p
          class="unnnic-text unnnic-text--size-body-gt unnnic-text--line-height-md unnnic-text--family-secondary unnnic-text--weight-bold unnnic-text--color-neutral-darkest"
          data-testid="title"
          data-v-73104bd4=""
          data-v-ec8da4ad=""
        >
          
          Test Title
          
        </p>
        <!--v-if-->
        <p
          class="unnnic-text unnnic-text--size-body-gt unnnic-text--line-height-md unnnic-text--family-secondary unnnic-text--weight-regular unnnic-text--color-neutral-cloudy header__description header__description"
          data-testid="description"
          data-v-73104bd4=""
          data-v-ec8da4ad=""
          title="Test Description"
        >
          
          Test Description
          
        </p>
      </header>
      <section
        class="content__skills"
        data-v-73104bd4=""
      >
        
        <section
          class="skill"
          data-testid="skill"
          data-v-29c532b5=""
          data-v-73104bd4=""
        >
          <p
            class="skill__icon"
            data-testid="skill-icon"
            data-v-29c532b5=""
          >
            icon-1
          </p>
          <p
            class="unnnic-text unnnic-text--size-body-md unnnic-text--line-height-md unnnic-text--family-secondary unnnic-text--weight-regular unnnic-text--color-neutral-cloudy skill__title skill__title"
            data-testid="skill-name"
            data-v-29c532b5=""
            data-v-ec8da4ad=""
          >
            
            Skill 1
            
          </p>
        </section>
        <section
          class="skill"
          data-testid="skill"
          data-v-29c532b5=""
          data-v-73104bd4=""
        >
          <p
            class="skill__icon"
            data-testid="skill-icon"
            data-v-29c532b5=""
          >
            icon-2
          </p>
          <p
            class="unnnic-text unnnic-text--size-body-md unnnic-text--line-height-md unnnic-text--family-secondary unnnic-text--weight-regular unnnic-text--color-neutral-cloudy skill__title skill__title"
            data-testid="skill-name"
            data-v-29c532b5=""
            data-v-ec8da4ad=""
          >
            
            Skill 2
            
          </p>
        </section>
        
      </section>
    </section>
    <button
      class="assign-agent-card__button unnnic-button unnnic-button--size-small unnnic-button--primary assign-agent-card__button"
      data-testid="assign-button"
      data-v-73104bd4=""
      data-v-d127031e=""
    >
      <!---->
      <!---->
      <!---->
      <span
        class="unnnic-button__label"
        data-testid="button-label"
        data-v-d127031e=""
      >
        
        
         Assign to the team
      </span>
      <!---->
    </button>
  </section>
  <assign-agent-drawer-stub
    agent="[object Object]"
    data-v-73104bd4=""
    isassigning="false"
    modelvalue="false"
  />
  
</div>
`;
